### PR TITLE
[PropertyInfo] Fix interface handling in `PhpStanTypeHelper`

### PR DESCRIPTION
**Pull request by janedbal**
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| License       | MIT

Previously, deserializing object with `Interface<T>` property was failing on undefined array key in `PhpStanTypeHelper` (generic `Clazz<T>` works just fine):

```
ErrorException: Undefined array key 0
/app/backend/vendor/symfony/property-info/Util/PhpStanTypeHelper.php:171
/app/backend/vendor/symfony/property-info/Util/PhpStanTypeHelper.php:49
/app/backend/vendor/symfony/property-info/Extractor/PhpStanExtractor.php:130
/app/backend/vendor/symfony/property-info/PropertyInfoExtractor.php:93
/app/backend/vendor/symfony/property-info/PropertyInfoExtractor.php:66
/app/backend/vendor/symfony/property-info/PropertyInfoCacheExtractor.php:102
/app/backend/vendor/symfony/property-info/PropertyInfoCacheExtractor.php:69
/app/backend/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:669
/app/backend/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:646
/app/backend/vendor/symfony/serializer/Normalizer/AbstractNormalizer.php:386
/app/backend/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:243
/app/backend/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:349
/app/backend/vendor/symfony/serializer/Serializer.php:247
/app/backend/vendor/symfony/serializer/Serializer.php:152
```

### CODE REVIEW

**Comment by janedbal** [comment]
Actually, this fix seems to be valid even for 6.4, should I rebase?

File: src/Symfony/Component/PropertyInfo/Util/PhpStanTypeHelper.php
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyInfo/Util/PhpStanTypeHelper.php]:
```suggestion
            if (!$collection && !class_exists($mainType->getClassName()) && !interface_exists($mainType->getClassName(), false)) {
```

File: src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpStanExtractorTest.php
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpStanExtractorTest.php]:
```suggestion
                    class: \BackedEnum::class,
```

File: src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpStanExtractorTest.php
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpStanExtractorTest.php]:
```suggestion
```

**Review by nicolas-grekas — COMMENTED** [review]
Please submit another PR for 6.4: this will help with merging up.

File: src/Symfony/Component/PropertyInfo/Util/PhpStanTypeHelper.php
Comment (janedbal) [reply src/Symfony/Component/PropertyInfo/Util/PhpStanTypeHelper.php]: Is this just microoptimization because `class_exists` already triggered autoloading and we dont need that for the second call of `interface_exists`?

File: src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpStanExtractorTest.php
Comment (janedbal) [reply src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpStanExtractorTest.php]: Shouldn't automated cs check in CI check that?

File: src/Symfony/Component/PropertyInfo/Util/PhpStanTypeHelper.php
Comment (stof) [reply src/Symfony/Component/PropertyInfo/Util/PhpStanTypeHelper.php]: it is indeed an optimization for the case where it does not exist, to avoid triggering autoloading twice.

File: src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpStanExtractorTest.php
Comment (janedbal) [reply src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpStanExtractorTest.php]: Anyway, adjusted.

File: src/Symfony/Component/PropertyInfo/Util/PhpStanTypeHelper.php
Comment (janedbal) [reply src/Symfony/Component/PropertyInfo/Util/PhpStanTypeHelper.php]: Anyway, adjusted.

Comment (nicolas-grekas) [reply src/Symfony/Component/PropertyInfo/Util/PhpStanTypeHelper.php]: That's correct

File: src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpStanExtractorTest.php
Comment (stof) [reply src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpStanExtractorTest.php]: @nicolas-grekas is fabbot using an uptodate version of PHP-CS-Fixer ? The `global_namespace_import` fixer should be enforcing that already (and it is part of the Symfony ruleset)

**Comment by janedbal** [comment]
> Please submit another PR for 6.4

Will do that once this is approved.

**Review by nicolas-grekas — APPROVED** [review]

**Review by mtarld — APPROVED** [review]

**Comment by xabbuh** [comment]
@janedbal Can you open the PR targeting 6.4?

**Comment by janedbal** [comment]
@xabbuh Sure, but it does not affect mergeability here, right?

**Comment by mtarld** [comment]
@janedbal, the 6.4 changes will be upmerged up to 7.3, therefore this 7.1 PR will become useless.

**Comment by janedbal** [comment]
Looks like we misunderstood each other, [so I just rebase it to 6.4](https://github.com/symfony/symfony/pull/59012#issuecomment-2503635525). No need [for another MR](https://github.com/symfony/symfony/pull/59012#pullrequestreview-2464779090), right?

**Comment by xabbuh** [comment]
If there is not much difference between the branches, rebasing the PR on 6.4 should be enough. Having a different PR for 7.1 would only be useful if there are conflicts to be expected when merging up which would then serve as a reference for us to ease resolving the conflicts.

**Comment by janedbal** [comment]
Rebased and changed base branch to `6.4`

**Comment by janedbal** [comment]
Is this mergeable now? Or are we waiting for code-owner approve? Or?

**Review by chalasr — APPROVED** [review]

**Comment by chalasr** [comment]
Thank you @janedbal.

**Comment by janedbal** [comment]
Thank you!

**Pull request merged**
